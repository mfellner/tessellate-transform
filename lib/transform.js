// @flow

import type { Parser, ParseResult } from './parsers';

export default function transform(
  content: string,
  parse: Parser<*>,
  root: ParseResult | null = null
): ParseResult {
  const nodes = root ? [root] : [];
  const getLastNode = () => nodes[nodes.length - 1];

  const callbacks = {
    onEnter: (node: ParseResult) => {
      const lastNode = getLastNode();

      if (typeof lastNode === 'object') {
        lastNode.children.push(node);
      } else {
        root = node;
      }
      nodes.push(node);
    },
    onLeave: () => {
      nodes.pop();
    },
    onLiteral: (literal: string) => {
      const lastNode = getLastNode();
      if (typeof lastNode === 'object') {
        lastNode.children.push(literal);
      }
    }
  };

  parse(content, callbacks);

  if (root && typeof root === 'object') return root;
  else throw new Error('Could not transform content.');
}
