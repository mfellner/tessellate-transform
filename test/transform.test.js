import fs from 'mz/fs';
import path from 'path';
import transform from '../lib/transform';
import createJSXParser from '../lib/parsers/jsx-parser';
import createJSONParser from '../lib/parsers/json-parser';

describe('transform', () => {
  it('should transform a simple JSX file', async () => {
    const buffer = await fs.readFile(path.resolve(__dirname, './fixtures/simple.jsx'));
    const result = transform(buffer.toString(), createJSXParser());
    expect(result).toMatchSnapshot();
  });

  it('should transform a JSX file with custom classes', async () => {
    const buffer = await fs.readFile(path.resolve(__dirname, './fixtures/custom.jsx'));
    const result = transform(buffer.toString(), createJSXParser({ typePrefix: 'my-prefix' }));
    expect(result).toMatchSnapshot();
  });

  it('should transform a simple JSON file', async () => {
    const buffer = await fs.readFile(path.resolve(__dirname, './fixtures/simple.json'));
    const result = transform(buffer.toString(), createJSONParser());
    expect(result).toMatchSnapshot();
  });

  it('should wrap a result in the provided root node', async () => {
    const root = {
      type: 'myroot',
      props: null,
      children: []
    };
    const buffer = await fs.readFile(path.resolve(__dirname, './fixtures/simple.json'));
    const result = transform(buffer.toString(), createJSONParser(), root);

    expect(typeof result).toBe('object');
    expect(result.type).toBe(root.type);
    expect(Array.isArray(result.children)).toBe(true);
  });

  it('transform an array of elements', async () => {
    const root = {
      type: 'myroot',
      props: null,
      children: []
    };
    const buffer = await fs.readFile(path.resolve(__dirname, './fixtures/array.json'));
    const result = transform(buffer.toString(), createJSONParser(), root);

    expect(typeof result).toBe('object');
    expect(result.type).toBe(root.type);
    expect(result).toMatchSnapshot();
  });
});
