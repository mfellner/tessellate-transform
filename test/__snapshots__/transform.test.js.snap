// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform should transform a JSX file with custom classes 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "props": Object {
                "alt": "Test",
                "src": "test.png",
              },
              "type": "img",
            },
          ],
          "props": Object {
            "sm": 6,
            "xs": 12,
          },
          "type": "my-prefix.Grid.Column",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "props": Object {
                "args": Object {
                  "bool": true,
                  "foo": "bar",
                },
              },
              "type": "my-prefix.Custom1",
            },
            Object {
              "children": Array [],
              "props": Object {
                "args": Object {
                  "test": 4.2,
                },
              },
              "type": "my-prefix.Custom2",
            },
          ],
          "props": Object {
            "sm": 6,
            "xs": 12,
          },
          "type": "my-prefix.Grid.Column",
        },
      ],
      "props": null,
      "type": "my-prefix.Grid.Row",
    },
  ],
  "props": null,
  "type": "my-prefix.Container",
}
`;

exports[`transform should transform a simple JSON file 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        "Hello, world!",
      ],
      "props": Object {
        "id": "heading",
      },
      "type": "h1",
    },
  ],
  "props": Object {
    "id": "test",
  },
  "type": "div",
}
`;

exports[`transform should transform a simple JSX file 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        "Hello, world!",
      ],
      "props": null,
      "type": "h1",
    },
  ],
  "props": Object {
    "id": "test",
  },
  "type": "div",
}
`;

exports[`transform transform an array of elements 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        "Hello, world!",
      ],
      "props": Object {
        "id": "heading1",
      },
      "type": "h1",
    },
    Object {
      "children": Array [
        "Another heading.",
      ],
      "props": Object {
        "id": "heading2",
      },
      "type": "h2",
    },
  ],
  "props": null,
  "type": "myroot",
}
`;
